presto-main/src/main/java/io/prestosql/operator/WorkProcessorOperatorAdapter.java

```diff
@@ -0,0 +1,201 @@
+                        pageBuffer);
+        this.pages = workProcessorOperator.getOutputPages();
+        this.memoryContext = operatorContext.localUserMemoryContext();
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public ListenableFuture<?> isBlocked()
+    {
+        if (!pages.isBlocked()) {
+            return NOT_BLOCKED;
+        }
+
+        return pages.getBlockedFuture();
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !isFinished() && !pageBuffer.isFinished() && pageBuffer.isEmpty();
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        pageBuffer.add(page);
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        if (!pages.process()) {
+            return null;
+        }
+
+        if (pages.isFinished()) {
+            return null;
+        }
+
+        return pages.getResult();
+    }
+
+    @Override
+    public void finish()
+    {
+        pageBuffer.finish();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return pages.isFinished();
+    }
+
+    @Override
+    public void close()
+            throws Exception
+    {
+        workProcessorOperator.close();
+    }
+
+    public class PageBuffer
+    {
+        @Nullable
+        private Page page;
+        private boolean finished;
+        @Nullable
+        private Runnable addPageListener;
+
+        public WorkProcessor<Page> pages()
+        {
+            return WorkProcessor.create(() -> {
+                if (!isEmpty()) {
+                    return ofResult(poll());
+                }
+
+                if (isFinished()) {
+                    return finished();
+                }
+
+                return yield();
+            });
+        }
+
+        @Nullable
+        public Page peek()
+        {
+            return page;
+        }
+
+        @Nullable
+        public Page poll()
+        {
+            Page page = this.page;
+            this.page = null;
+            memoryContext.setBytes(0);
+            return page;
+        }
+
+        public boolean isEmpty()
+        {
+            return page == null;
+        }
+
+        public boolean isFinished()
+        {
+            return finished;
+        }
+
+        public void setAddPageListener(Runnable stateChangeListener)
+        {
+            this.addPageListener = requireNonNull(stateChangeListener, "stateChangeListener is null");
+        }
+
+        private void add(Page page)
+        {
+            checkState(isEmpty());
+            this.page = requireNonNull(page, "page is null");
+
+            if (addPageListener != null) {
+                addPageListener.run();
+            }
+
+            // if page was not immediately consumed, account it's memory
+            if (!isEmpty()) {
+                memoryContext.setBytes(page.getSizeInBytes());
+            }
+        }
+
+        private void finish()
```

Comment:
I may have missed it, but I think there is only one caller of this, so I would inline this.  I think it might be clear to readers if this is not something to be called (since this is an inner class).

Reply:
> I think it might be clear to readers if this is not something to be called (since this is an inner class).

Isn't private keyword sufficient for that? This method is similar to `add` method directly above in this regard. Anyway in this case it seems cleaner that private `PageBuffer` field is not directly modified from wrapping class (as none of other fields).

Reply:
The class containing the inner class is allowed access to any method or field regardless of access modifier.